File: leetcode-python/_0xx/_120_triangle.py
import unittest
from typing import List

# Date: 2025-09-25
# Problem: 120 triangle
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        memo = {}
        def dp(r, c):
            if r == len(triangle) - 1:
                return triangle[r][c]
            if (r,c) in memo:
                return memo[(r, c)]
            res = min(triangle[r][c] + dp(r+1,c), triangle[r][c] + dp(r+1,c+1))
            memo[(r,c)] = res
            return res
        return dp(0, 0)


s = Solution()
class MyTestCase(unittest.TestCase):

    def test_case_1(self):
        actual = s.minimumTotal(triangle = [[2],[3,4],[6,5,7],[4,1,8,3]])
        self.assertEqual(11, actual)
        
    def test_case_2(self):
        actual = s.minimumTotal(triangle = [[-10]])
        self.assertEqual(-10, actual)

    def test_case_3(self):
        actual = s.minimumTotal(triangle = [[1], [2, 3]])
        self.assertEqual(3, actual)

if __name__ == '__main__':
    unittest.main()

File: leetcode-cpp2/_0xx/_120_triangle.cpp
#include <catch2/catch_test_macros.hpp>
#include "../utils/serde.cpp"
#include <map>
// Date: 2025-09-25
// Problem: 120 triangle
using std::vector;
using std::string;

class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        std::map<std::pair<int, int>, int> memo;
        std::function<int(int, int)> dp = [&triangle, &dp, &memo](int r, int c) -> int {
            if (r == triangle.size() - 1) {
                return triangle[r][c];
            }
            std::pair key = {r, c};
            if (memo.contains(key)) {
                return memo[{r, c}];
            }
            int down = dp(r+1, c);
            int diag = dp(r+1, c + 1);
            int res = triangle[r][c] + std::min(down, diag);
            memo[{r, c}] = res;
            return res;
        };

        return dp(0, 0);
    }
};

Solution s;
TEST_CASE("[120] Test case 1", "") {
    auto input = parseMatrix("[[2],[3,4],[6,5,7],[4,1,8,3]]");
    auto actual = s.minimumTotal(input);
    REQUIRE(actual == 11);
}


TEST_CASE("[120] Test case 2", "") {
    auto input = parseMatrix("[[-10]]");
    auto actual = s.minimumTotal(input);
    REQUIRE(actual == -10);
}

TEST_CASE("[120] Test case 3", "") {
    auto input = parseMatrix("[[1], [2, 3]");
    auto actual = s.minimumTotal(input);
    REQUIRE(actual == 3);
}
